tools/mksurfdata_esmf/src/mksoiltexMod.F90

```diff
@@ -98,229 +103,270 @@ subroutine mksoiltex(file_mesh_i, file_data_i, mesh_o, pioid_o, pctlnd_pft_o, rc
+
+    rcode = pio_openfile(pio_iosystem, pioid_i, pio_iotype, trim(file_lookup_i), pio_nowrite)
+
+    rcode = pio_inq_dimid  (pioid_i, 'MapUnit', dimid)
+    rcode = pio_inq_dimlen (pioid_i, dimid, n_mapunits)
+
+    rcode = pio_inq_dimid  (pioid_i, 'soil_layer', dimid)
     rcode = pio_inq_dimlen (pioid_i, dimid, nlay)
-    allocate(sand_i(mapunit_value_max,nlay), stat=ier)
+
+    rcode = pio_inq_dimid  (pioid_i, 'SCID', dimid)
+    rcode = pio_inq_dimlen (pioid_i, dimid, n_scid)
+
+    ! Read In MapUnits from the input file
+    allocate(MapUnits(n_mapunits), stat=ier)
+    rcode = pio_inq_varid(pioid_i, 'MapUnit', pio_varid)
+    rcode = pio_get_var(pioid_i, pio_varid, MapUnits)
+
+    ! Determine the mapunit lookup index from the value of the MapUnit variable
+    mapunit_value_max = maxval(MapUnits)
+    allocate(mapunit_lookup(mapunit_value_max))
+    mapunit_lookup(:) = -999
+    do n = 1,size(MapUnits)
+       mapunit_lookup(MapUnits(n)) = n
+    end do
+
+    allocate(sand_i(nlay,n_scid,n_mapunits))
+    if (ier/=0) call shr_sys_abort()
+    allocate(clay_i(nlay,n_scid,n_mapunits))
     if (ier/=0) call shr_sys_abort()
-    allocate(clay_i(mapunit_value_max,nlay), stat=ier)
+    allocate(orgc_i(nlay,n_scid,n_mapunits))
     if (ier/=0) call shr_sys_abort()
 
-    ! read in sand_i and clay_i (they will read in total on all processors)
-    rcode = pio_inq_varid(pioid_i, 'PCT_SAND', pio_varid)
-    rcode = pio_get_var(pioid_i, pio_varid, sand_i)
-    rcode = pio_inq_varid(pioid_i, 'PCT_CLAY', pio_varid)
-    rcode = pio_get_var(pioid_i, pio_varid, clay_i)
+    ! Get dimensions from input file and allocate memory for sand_i, clay_i,
+    ! and organic carbon content
+    rcode = pio_inq_varid(pioid_i, 'PCT_SAND', pio_varid_sand)
+    rcode = pio_inq_varid(pioid_i, 'PCT_CLAY', pio_varid_clay)
+    rcode = pio_inq_varid(pioid_i, 'ORGC', pio_varid_orgc)
 
-    ! Set soil texture as follows:
-    ! a. Use dominant igbp soil mapunit based on area of overlap unless 'no data' is dominant
-    ! b. If this has no data, use loam for soil texture
+    rcode = pio_get_var(pioid_i, pio_varid_sand, sand_i)
+    rcode = pio_get_var(pioid_i, pio_varid_clay, clay_i)
+    rcode = pio_get_var(pioid_i, pio_varid_orgc, orgc_i)
 
     do no = 1,ns_o
-       if (mapunit_o(no) > 0) then
-          ! valid value is obtained
-          if (mapunit_o(no) > mapunit_value_max) then
-             write(6,*)'mapunit_o is out of bounds ',mapunit_o(no)
-             ! call shr_sys_abort("mapunit_o is out of bounds")
-          end if
-          do l = 1, nlay
-             sand_o(no,l) = sand_i(mapunit_o(no),l)
-             clay_o(no,l) = clay_i(mapunit_o(no),l)
-          end do
+
+       if (pctlnd_pft_o(no) < 1.e-6_r8 .or. mapunit_o(no) == 0) then
+
+          ! Adjust sand and clay be loam if pctlnd_pft is < 1.e-6 or mapunit is 0
+          sand_o(no,:) = 43._r4
+          clay_o(no,:) = 18._r4
+          orgc_o(no,:) = 18._r4  ! TODO slevis: value for loam here?
```

Comment:
@wwieder these are the model's generic values for sand and clay. What do we want for organic in that case?

Also, I haven't included a units transformation here from gC/kg (lookup file) to kg/m3 (fsurdat), which I think is what you were discussing in #1303. Did you reach a verdict?

Reply:
Sam what are the organic values currently used in oceans?  I'm included to do the same here?

I'm not sure we ever decided how to handle units, let's set up a meeting with @dlawrenncar to discuss after the tutorial.  I'll try setting up a meeting.

Reply:
> Sam what are the organic values currently used in oceans? I'm included to do the same here?

Sounds good. So I changed to 0 here and also in cases when the lookup table returns a negative value. But not when the lookup table returns -4 (sand dunes) where I set organic = 1. This is on the low side compared to what I see from our old dataset. We can address this, too, in our meeting.